### DISCUSSION TITLE
Allow setting alternative destination in SAML configuration

### DISCUSSION THREAD

**Discussion by richardalberto** [keycloak/keycloak#22350] [Ideas]
## The problem
Our customers have created their SAML connections with a legacy metadata fields before we started adopting Keycloak; the destination field is set on the IdP side to our legacy URL which today serves as our SAML processing endpoint. 

Keycloak has, and expects, per-identity-provider connection brokering endpoints in the format of `https://keycloack.instance/realms/<realm>/broker/<endpoint-entity-id>/endpoint`. Therefore, if we were to simply proxy the legacy endpoint to Keycloak, Keycloak will reject the request, as it verifies that the destination field matches the URL path it is expecting. 

Since the destination field is configured within the customers external IdP it cannot be updated programmatically and would require customer intervention.

## Possible solutions we've discussed
- **Creating a proxy that intercepts the requests and mutates the destination url**: This comes with it's own set of complexities as we would need to mutate fields in the request and re-sign the request with a certificate we trust. Even if we work thru all the complexities of this approach; we still have to maintain an additional service.
- **Vendor/Change keycloak validation and allow the legacy URL in code**: This is doable but is something that we are trying to avoid if possible since again it's adding complexity to our CI/CD process and makes upgrades harder in the future. 

## 

This feels like a common use case for companies adopting keycloak and might benefit others going thru a similar situation. I'm curious if this is something that would benefit the product and more specifically I have a few questions:

- Is this something the should be part of keycloak? 
- Is there any security concerns with a feature like this one?

**Comment by kristofferlanno** [comment]
Hi @richardalberto ! We're running into the same issue when trying to implement SAML brokering for our OIDC client. We can easily get SP-initiated to work with using Keycloak's default endpoint (the one you posted in your message), but getting idp-initiated to work is proving to be not as straight forward :/ An idp-initiated assertion towards that endpoint throws an error due to RelayState being null. We've tried creating a separate SAML client and using the `/endpoint/clients/<client-url-name>` method. This authenticates the user in Keycloak and, while a bit ugly, allows us to redirect to the endpoint in our app that starts the OIDC flow (which logs in the user right away since they are now already logged in with Keycloak).

The issue we run into is that the SP-initiated and IdP-initiated flows now have two different ACS URLs. SP uses `/endpoint` while IdP uses `/endpoint/clients/<client-url-name>`. We have tried creating a new endpoint to serve as ACS URL which would detect what flow is used and do a `307 Temporary Redirect` towards the right endpoint. But here is of course where we run into the issue you mention with Keycloak rejecting the request due to a mismatching Destination. 

There doesn't seem to be a way in Keycloak to specify the destination we expect. Did you manage to solve your problem somehow?

**Comment by kristofferlanno** [comment]
Frankly, if we could avoid the whole SAML Client thing and just have a way of specifying a "default client" or a URL to redirect to, after Keycloak has verified the assertion and created a session for the user, on our external IdP configuration then that would simplify a lot for us. We could then simply specify our OIDC apps login path as a redirect URL.

**Comment by navinkaushik** [comment]
@richardalberto We are facing exactly the same issue. Previously, we were using saml between idp & application, now we want to move from application to Keycloak and we don't want to change redirect url in idp which is in customer environment. We forward the saml response coming to old url to Keycloak but Keycloak gives an invalid request error, probably due to destination validation as old redirect url in destination field in saml response.  Any idea, how this can be resolved ?
